crates/ra_hir_ty/src/traits.rs

```diff
@@ -195,7 +221,25 @@ pub(crate) fn trait_solve_query(
     // We currently don't deal with universes (I think / hope they're not yet
     // relevant for our use cases?)
     let u_canonical = chalk_ir::UCanonical { canonical, universes: 1 };
-    let solution = db.trait_solver(krate).solve(db, &u_canonical);
+
+    let solution = {
+        let solver = db.trait_solver(krate);
+
+        // chalk sometimes panics, so let's treat this as a no-solution
+        // situation.
+        let db = panic::AssertUnwindSafe(db);
+        match panic::catch_unwind(|| solver.solve(db.0, &u_canonical)) {
```

Comment:
hmm, I'm confused, why are we doing the panic handling twice? isn't this just calling the `solve` method above which already catches panics?

Reply:
`match self.inner.lock() ` handles poisoning, `panic::catch_unwind` handles panics.

Overall, there are four cases here:

* happy path: everyting works, no cancellation happens
* cancellation: `solution` is `Err(cancelled)`, and we need to re-raise this panic (`panic::resume_unwind(err)`)
* an already cancelled request: we try to lock chalk, but the mutex is poisoned. It is poisoned because some other thread got cancelled, and unwound past the mutex guard. Here, we need to initiate unwinding on the current thread `Err(_) => ra_db::Canceled::throw()`
* chalk panicked. We need to reset the solver `*solver = create_chalk_solver();`

At least, I *think* this is what happens here, I've re-written this like five times, putting `catch_unwind` in different places. One invariant we have here is that if the mutex is poisoned, it is poisoned because of the cancellation.

Reply:
Ok, but there's two `catch_unwind`s -- the one here and the one above in `TraitSolver::solve`, which is just what we're calling here, isn't it?

Reply:
:man_facepalming: that's... actually the remains of one of those five attempts.

Excellent catch!